libnymea-core/devices/scriptdeviceplugin.cpp

```diff
@@ -0,0 +1,255 @@
+ *                                                                         *
+ *  Copyright (C) 2019 Michael Zanetti <michael.zanetti@nymea.io>          *
+ *                                                                         *
+ *  This file is part of nymea.                                            *
+ *                                                                         *
+ *  nymea is free software: you can redistribute it and/or modify          *
+ *  it under the terms of the GNU General Public License as published by   *
+ *  the Free Software Foundation, version 2 of the License.                *
+ *                                                                         *
+ *  nymea is distributed in the hope that it will be useful,               *
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of         *
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the          *
+ *  GNU General Public License for more details.                           *
+ *                                                                         *
+ *  You should have received a copy of the GNU General Public License      *
+ *  along with nymea. If not, see <http://www.gnu.org/licenses/>.          *
+ *                                                                         *
+ * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+#include "scriptdeviceplugin.h"
+
+#include <QQmlEngine>
+#include <QDir>
+#include <QJsonDocument>
+
+#include "loggingcategories.h"
+#include <plugintimer.h>
+
+ScriptDevicePlugin::ScriptDevicePlugin(QObject *parent) : DevicePlugin(parent)
+{
+
+}
+
+bool ScriptDevicePlugin::loadScript(const QString &fileName)
+{
+
+    QFileInfo fi(fileName);
+    QString metaDataFileName = fi.absoluteDir().path() + '/' + fi.baseName() + ".json";
+
+    QFile metaDataFile(metaDataFileName);
+    if (!metaDataFile.open(QFile::ReadOnly)) {
+        qCWarning(dcDeviceManager()) << "Failed to open plugin metadata at" << metaDataFileName;
+        return false;
+    }
+    QJsonParseError error;
+    QByteArray data = metaDataFile.readAll();
+    QJsonDocument jsonDoc = QJsonDocument::fromJson(data, &error);
+    metaDataFile.close();
+
+    if (error.error != QJsonParseError::NoError) {
+        int errorOffset = error.offset;
+        int newLineIndex = data.indexOf("\n");
+        int lineIndex = 1;
+        while (newLineIndex > 0 && errorOffset > newLineIndex) {
+            data.remove(0, newLineIndex + 2);
+            errorOffset -= (newLineIndex + 2);
+            newLineIndex = data.indexOf("\n");
+            lineIndex++;
+        }
+        if (newLineIndex >= 0) {
+            data = data.left(newLineIndex);
+        }
+        QString spacer;
+        for (int i = 0; i < errorOffset; i++) {
+            spacer += ' ';
+        }
+        QDebug dbg = qWarning(dcDeviceManager()).nospace().noquote();
```

Comment:
Now this is interesting. I was just testing if this actually worked, because I had vague memories of `qDebug(QLoggingCategory)` not even compiling. And indeed, it works as expected (the message is in the correct logging category).

So I started to dig the underlying code and guess what, since Qt 5.3 (!), both work:
`qCWarning(QLoggingCateogory)`
as well as `qWarning(QLoggingCategory)`

qWarning actually expands to

`QMessageLogger(QT_MESSAGELOG_FILE, QT_MESSAGELOG_LINE, QT_MESSAGELOG_FUNC).warning` 
and there is a QMessageLogger::warning() variant which takes a QLoggingCategory as argument.

So this means, we can also just use `qDebug, qWarning` with logging categories. The only difference is really that qCWarning requires an argument while qWarning does not require it.

I'd probably vote for keep going with the `C` variants as they force the usage of a category and make it less likely to mistakenly use it w/o a category.

However, in this particular case, where I need the QDebug object to compose a multi-line warning, `qCWarning` doesn't do... (that only expands to a for-loop) so I need to use qWarning(), or create a QLoggingCategory manually.

So I guess we're fine in this particular case...